### ISSUE TITLE
Support Orphans & Widows

### ISSUE THREAD

**Issue by espadrine** [johnfactotum/foliate#644]
Hi! I love this app.

One classic typesetting feature that I am used to, and whose absence leads to a tiny decrease in readability, is the avoidance of [orphans and widows][WP].

**Orphans and widows** are lines at the start or end of paragraphs, that end up alone because of pagination cuts. That removes their context, which undermines the reading flow.

![foliate-orphan-widow](https://user-images.githubusercontent.com/100689/102676897-03d81e00-41a0-11eb-94bd-f4ec12fdefda.png)

*(Example orphan and widow. Well, the orphan shown is not technically one, but it is a related issue.)*

A common solution is *to move the orphan to the next page*, or to move one line over to the widow’s page so it is not alone.

[CSS][W3C] [has a solution][MDN] which [is implemented][CIU] in Chrome, Safari (and IE 10+, although I doubt that adds any relevance).

For good measure, I tried adding it to the body:

![image](https://user-images.githubusercontent.com/100689/102677484-f4a69f80-41a2-11eb-9dc1-b2a6cbc969ec.png)

…and the result looks satisfactory!

![image](https://user-images.githubusercontent.com/100689/102677535-346d8700-41a3-11eb-81e1-256a146dba4d.png)

[WP]: https://en.wikipedia.org/wiki/Widows_and_orphans
[W3C]: https://www.w3.org/TR/CSS2/page.html#break-inside
[MDN]: https://developer.mozilla.org/en-US/docs/Web/CSS/orphans
[CIU]: https://caniuse.com/?search=orphan

**Comment by johnfactotum** [comment]
One thing about setting orphans and widows is that it produces a ragged bottom, which is sometimes (or at least to some people) more undesirable than orphans and widows. I guess it's always possible to (a) make it optional, or (b) compromise by allowing widows but not orphans, or (c) enable this only when text is not justified.

> Well, the orphan shown is not technically one, but it is a related issue.

Yeah, and this problem is actually much worse than orphans. Unfortunately, I don't think it's possible to fix this, because WebKit does not support [`break-after` in multi-column layouts](https://caniuse.com/mdn-css_properties_break-after_multicol_context).

**Comment by J-J-Chiarella** [comment]
@johnfactotum 

> One thing about setting orphans and widows is that it produces a ragged bottom, which is sometimes (or at least to some people) more undesirable than orphans and widows.

I doubt this being more undesirable or at least I doubt it being more *widely* undesirable. I cannot think of a single periodical (as in an academic journal, not a typical newspaper with its super slim columns and helter-skelter formatting) or book that prioritized a justified bottom to the point of tolerating widows and orphans. It's impossible to do if you ever have sections or diagrams. At some point, you would get a situation as shown in the OP's image. The spacing above/below or different font sizes mean that a perfectly justified vertical layout is a non-starter.

Vertically justified text could work even if you handled widows and orphans "properly." You just make the inter-line spacing adjust so that the text always took up 90% (or whatever percent) of the screen. I've fiddled with this in various ways, even on paper where I would draw lines on blank paper and write out a bunch of text to see what it looked like. I had to burn the paper, my office, and the building, then jam out my eyeballs. There is a reason that no one publishes this way.

Widows and orphans in any script make me think of vanity published "books" and church newsletters and middle school newspaper clubs.

> I guess it's always possible to (a) make it optional, or (b) compromise by allowing orphans but not widows, or (c) enable this only when text is not justified.

(A) and (c) have their respective pros and cons. Tough choice. I can't see (b) as a compromise so much as I see it as a half-way implemented solution, however.

> Yeah, and this problem is actually much worse than orphans. Unfortunately, I don't think it's possible to fix this, because WebKit does not support [`break-after` in multi-column layouts](https://caniuse.com/mdn-css_properties_break-after_multicol_context).

That's too bad to hear. That example you and the OP are referring to ... that's hideous. Is the "break-after" issue something that one would report to WebKit or GTK-Webkit or ...?

**Comment by johnfactotum** [comment]
That may be so. But it's not about having the text justified perfectly, but more about the overall balance. I did try setting orphans and widows some while ago, and found that the unbalance is often very distracting. Maybe it's just my personal preference or maybe I'm just used to it.

Note that the publisher can always set the CSS themselves if they want. A lot of them are already doing that. So it's not a real problem for EPUBs and Kindle files.

> I can't see (b) as a compromise so much as I see it as a half-way implemented solution, however.

Surely, a widow (or worse, a lone widow at the end of a chapter) is much more egregious? According to the latest _Chicago Manual of Style_ (17 edition, section 2.116):

> A page should not end with a subhead. Nor should a page begin with the last line of a paragraph unless it is full measure; a short line in this position is some­ times called a widow. **A page can, however, end with the first line of a new paragraph, or what is sometimes referred to as an orphan** [emphasis added].

So in my mind this is probably a good solution.

> Is the "break-after" issue something that one would report to WebKit or GTK-Webkit or ...?

Actually they do support `-webkit-column-break-after` and `-webkit-column-break-before`, however these don't support `avoid` so they are useless for this issue.

I believe this is the bug report: https://bugs.webkit.org/show_bug.cgi?id=61775. I should note that, in general, the multicol support in WebKit is quite poor. There's no support for nested columns, and the performance is poor compared to other browsers, so there's definitely a
lot of rooms for improvement in that area.

**Comment by J-J-Chiarella** [comment]
Usually I side with Chicago style, but not here. I violate a couple of small things that Chicago suggests, as do most writers and publishers who claim to use Chicago. It explicitly says to use spaced periods (regular, non-hair spaces) and not dedicated ellipsis characters, but I always use ellipses. I use LaTeX and have used office word processors before and the spaced periods never match what I see in publications for most fonts like Times New Roman.

I really don't understand Chicago's reasoning on this at all. Imagine you are writing something on a paper. The text takes up slightly more than a page. You flip the page over to continue writing and write about half of a line more. It wouldn't upset me as a reader, if I turn the page over and see only one line. On the other hand, if I am reading something is multiple pages long, I would rather not see orphans. If I am writing something, near the end of a page, and know that I have a lot more to write, then I will turn the page over when I finish my paragraph even if exactly one line remains for me to start an orphan line for the new paragraph.

I wouldn't be surprised if Chicago flip-flopped on this or does in the future. The treatment of internet sources has been all over the map for the past 15 years with Chicago.

When you wrote

> Maybe it's just my personal preference or maybe I'm just used to it.

I wonder if your being used to it is largely consuming books on e-readers and screens. (Do you?) I am very loath to do so most times. Because most of my reading is on dead trees, I am very used to seeing the convention of preventing widows and orphans.

> ...  (b) compromise by allowing orphans but not widows, which is sometimes (or at least to some people) more undesirable than orphans and widows ...

I doubted the utility of (b) because I didn't see it as a compromise with anyone. I thought you were worried about some group of users that I believed to simply not exist. I can now infer you were counting yourself in the group of those not so gung ho on widows and orphans, whereas I say to nuke them from orbit.

Anyhow, thanks for the detail here and your regular work on the program! Linux sorely needed a good reader besides Calibre. Using that program to read can feel like firing up Kodi or Blender to watch a thirty second video clip you downloaded.

**Comment by johnfactotum** [comment]
The reasoning of the _Chicago Manual of Style_ is clear, I believe, from the heading of the section, which is "Proofreading for overall appearance". Orphans have far less impact on the overall appearance compared to widows.

To quote [_Butterick’s Practical Typography_](https://practicaltypography.com/widow-and-orphan-control.html),

> Widow and orphan control in a word processor is all-or-nothing. You can’t control widows and orphans separately, even though widows are more distracting. Why? Orphans appear at the beginning of a paragraph, so they’re at least a full line. But widows can be any length, even a single word, because they appear at the end of a paragraph. [...]
> In my own work, I approach widow and orphan control the same way I approach ligatures—I only use it if widows and orphans are causing a visible problem. Otherwise, I find that the blank lines at the bottom of the page are more annoying than the widows and orphans.

When both orphans and widows are prevented, it's not uncommon or unreasonable, I think, to require more lines for widows than orphans (e.g. two lines for orphans, but three lines for widows) for the same reason.

> I thought you were worried about some group of users that I believed to simply not exist.

Well, I don't know about Foliate users, but I do see some people (on the internet) who do prefer to have a less ragged bottom over preventing widows and orphans.

**Comment by johnfactotum** [comment]
The gtk4 branch now sets `widows: 2` and nothing about orphans. I think this is a good enough compromise for now, as I don't really want add more options. But we can revisit this later if there's enough demand.

**Comment by J-J-Chiarella** [comment]
> The gtk4 branch now sets `widows: 2` and nothing about orphans. I think this is a good enough compromise for now, as I don't really want add more options. But we can revisit this later if there's enough demand.

First, thank you for addition and the GTK4 port work. I appreciate your wanting to see about wider demand.

What I write below is for posterity, not to fight the decision.

> Well, I don't know about Foliate users, but I do see some people (on the internet) who do prefer to have a less ragged bottom over preventing widows and orphans.

As I have gotten more experience with LaTeX (and gotten the spacing to work to my preferences, and am now doing the Chicago recommendation of spaced periods instead of ellipsis characters, which can be awful in some typefaces), I have also been paying more attention to line spacing.

I think (or thought) that the distaste for widows, for orphans, and for ragged bottoms had been solved long ago by typographers. The orthodoxy had been: (1) Prevent all widow lines and orphan lines. The minimum is always two. (2) Justify the inter-line spacing of shorter columns so that the tops and bottoms of all full columns are flush with one another.

Websites seem to be able to handle left-to-right and right-to-left justification fine, but they seem to fail on word hyphenation, widows, orphans, and inter-line spacing. Hence, I think that the tolerance for widows and orphans for the sake of avoiding  ragged bottoms is a relatively new phenomenon. It is as if the proliferation of web resources has sent typographical conventions back to the seventeenth century or something. I cannot understand why we (humanity) should take a step backwards in time in order to accommodate new technology. Usually, the modus operandi should be to implement all of the old features in the new technologies.

**But ...**

as I said, I can respect the currently working decision.

With writing a dissertation that is required to have no widows or orphan lines (good) but must have no horizontal justification (?!?) and no inter-line spacing (?!?) and no footnotes spilling over onto following pages, I have grown to abhor the large white spaces, but I also question why we cannot just have automated inter-line spacing, but some of the gaps are so large (due to graphs or footnotes), that inter-line spacing would have to be ridiculous.

Okay ... so we don't mimic typewriters' monospaced font and we let computers take care of moving words around so there is no need for word-splitting or hyphenation, but we pretend to be using typewriters (ragged sides and bottoms) ... and also maintain double-spacing, which can make the problems from justifying the inter-line spacing even greater.

TL;DR: I can see situations in which naively preventing orphans will cause more harm than good.

Thank you for work.
